src/nvim/os/fs.c

```diff
@@ -286,35 +282,55 @@ static bool is_executable(const char *name)
 #ifdef WIN32
   // Windows does not have exec bit; just check if the file exists and is not
   // a directory.
-  return (S_ISREG(mode));
+  const bool ok = S_ISREG(mode);
 #else
   int r = -1;
   if (S_ISREG(mode)) {
     RUN_UV_FS_FUNC(r, uv_fs_access, name, X_OK, NULL);
   }
-  return (r == 0);
+  const bool ok = (r == 0);
 #endif
+  if (ok && abspath != NULL) {
+    *abspath = save_abs_path((char_u *)name);
+  }
+  return ok;
 }
 
 #ifdef WIN32
-/// Appends file extensions from `pathext` to `name` and returns true if any
-/// such combination is executable.
-static bool is_executable_ext(char *name, const char *pathext)
-  FUNC_ATTR_NONNULL_ALL
+/// Checks if file `name` is executable under one of these conditions:
+/// - if the file extension is in $PATHEXT
+/// - if the result of any $PATHEXT extension appended to `name` is executable
+static bool is_executable_ext(char *name, char_u **abspath)
+  FUNC_ATTR_NONNULL_ARG(1, 2)
 {
+  char *nameext = strrchr(name, '.');
+  size_t nameext_len = nameext ? strlen(nameext) : 0;
   xstrlcpy(os_buf, name, sizeof(os_buf));
   char *buf_end = xstrchrnul(os_buf, '\0');
+  const char *pathext = os_getenv("PATHEXT");
+  if (!pathext) {
+    pathext = ".com;.exe;.bat;.cmd";
+  }
   for (const char *ext = pathext; *ext; ext++) {
-    // Skip the extension if there is no suffix after a '.'.
+    // If $PATHEXT itself contains dot:
     if (ext[0] == '.' && (ext[1] == '\0' || ext[1] == ENV_SEPCHAR)) {
+      if (is_executable(name, abspath)) {
+        return true;
+      }
+      // Skip the extension.
       ext++;
       continue;
     }
 
     const char *ext_end = xstrchrnul(ext, ENV_SEPCHAR);
-    STRLCPY(buf_end, ext, ext_end - ext + 1);
-
-    if (is_executable(os_buf)) {
+    size_t ext_len = (size_t)(ext_end - ext);
+    STRLCPY(buf_end, ext, ext_len + 1);
+    bool in_pathext = (strstr((char *)path_tail(p_sh), "sh") != NULL)
```

Comment:
could move the "sh" check outside the loop:

    bool is_unix_shell = strstr((char *)path_tail(p_sh), "sh") != NULL;